Mathlib/NumberTheory/LucasPrimality.lean

```diff
@@ -59,3 +58,74 @@ theorem lucas_primality (p : ℕ) (a : ZMod p) (ha : a ^ (p - 1) = 1)
     p - 1 = orderOf a := order_of_a.symm
     _ = orderOf a' := (orderOf_injective (Units.coeHom (ZMod p)) Units.ext a')
     _ ≤ Fintype.card (ZMod p)ˣ := orderOf_le_card_univ
+
+
+/-- If `p` is prime, then there exists an `a`
+such that `a^(p-1) = 1 mod p` and `a^((p-1)/q) ≠ 1 mod p` for all prime factors `q` of `p-1`,
+The multiplicative group mod `p` is cyclic, so `a` can be any generator of the group
+(which must have order `p-1`)
+-/
+theorem reverse_lucas_primality (p : ℕ) (hP : p.Prime):
+    ∃ a : ZMod p, a ^ (p - 1) = 1 ∧ ∀ q : ℕ, q.Prime → q ∣ p - 1 → a ^ ((p - 1) / q) ≠ 1 := by
+
+  have _: Fact p.Prime := ⟨hP⟩
+  have _: 1 < p := by exact Nat.Prime.one_lt hP
+
+  have h0: IsCyclic ((ZMod p)ˣ) := by exact instIsCyclicUnitsOfFinite
+  obtain ⟨g, hg⟩ := h0.exists_generator
+
+  have h1: orderOf g = p - 1 := by
+    rw [orderOf_generator_eq_natCard, Nat.card_eq_fintype_card, ← Nat.prime_iff_card_units]
+    trivial
+    trivial
+
+  -- Our generator 'g' works as the witness 'a'
+  use g
+  rw [
+    ← (orderOf_injective (Units.coeHom (ZMod p)) Units.ext _),
+    Units.coeHom_apply,
+    orderOf_eq_iff
+  ] at h1
+
+  constructor
+
+  exact h1.1
+
+  -- Show that 'g^((p-1)/q) ≠ 1', using the fact that '(p-1)/q)' < 'orderOf g'
+  intro q hq hq_div
+  have _: q ≤ p - 1 := by
+    apply Nat.le_of_dvd at hq_div
+    exact hq_div
+    simp only [tsub_pos_iff_lt]
+    trivial
+  have q_gt_1: 1 < q := by
+    apply Nat.Prime.one_lt
+    exact hq
+  have p_div_lt: (p - 1) / q < p - 1 := by
+    apply Nat.div_lt_self
+    simp only [tsub_pos_iff_lt]
+    linarith
+    exact q_gt_1
+  have neq_one: g.val ^ ((p - 1) / q) ≠ 1 := by
+    apply h1.2
+    exact p_div_lt
+    apply Nat.div_pos
+    linarith
+    linarith
+
+  exact neq_one
+  simp
+  linarith
```

Comment:
Here's a condensed version of your proof:
```suggestion
  have : Fact p.Prime := ⟨hP⟩
  obtain ⟨g, hg⟩ := IsCyclic.exists_generator (α := (ZMod p)ˣ)
  have h1 : orderOf g = p - 1 := by
    rwa [orderOf_generator_eq_natCard hg, Nat.card_eq_fintype_card, ← Nat.prime_iff_card_units]
  have h2 := tsub_pos_iff_lt.2 hP.one_lt
  rw [← orderOf_injective (Units.coeHom _) Units.ext _, orderOf_eq_iff h2] at h1
  refine ⟨g, h1.1, fun q hq hqd ↦ ?_⟩
  replace hq := hq.one_lt
  exact h1.2 _ (Nat.div_lt_self h2 hq) (Nat.div_pos (Nat.le_of_dvd h2 hqd) (zero_lt_one.trans hq))
```
I removed the comments, since I don't think they're terribly valuable now that the proof is relatively short (and since the docstring already explains the proof anyways). But feel free to add them back if you think they still provide value.

Reply:
Here's some tips that might be helpful in the future:
- `linarith` is a very powerful tactic, and for that reason, it should be reserved for whenever it's absolutely necessary. Almost all uses in your proof were equivalent to `assumption` (and the remaining one was just `1 < q → 0 < q`, so still overkill).
- There usually isn't much reason to use `trivial`, especially when you can replace it by `assumption` or `exact`.
- When you have multiple goals, you should isolate each using `·` and an indent instead of solving them one after the other.

Reply:
Thanks for the help - this is much cleaner than what I came up with. I updated your PR to use your version of the proof, and made another simplification from my original (`orderOf_generator_eq_natCard, Nat.card_eq_fintype_card` can be replaced with `orderOf_eq_card_of_forall_mem_zpowers` ).

Since this is such an extensive simplification, I added you as a co-author